internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala

```diff
@@ -59,7 +59,12 @@ object Locate {
     def entries = classpath.iterator.map { entry =>
       (entry, lookup.definesClass(entry))
     }
-    className => entries.collectFirst { case (entry, defines) if defines(className) => entry }
+    className =>
+      // See sbt/zinc#757, sbt/zinc#925. Class name containing "<" is usually a synthetic
+      // one that does not have a corresponding *.class file.
```

Comment:
What?  Classes might be synthetic, but they still have classfiles, unless they're lambdas.

This looks like a hackaround without understanding the original issue and where the "<" isn't being encoded.

Reply:
Note this is different from type name. _Before_ someone asks `Locate` to locate the corresponding `*.class` lookup they have to figure out the mapping between the type and the class, and it will not be obvious from just the type name. Here's an example:

```scala
package example

object Foo extends App {
  trait Greeting {
    lazy val greeting: String = "hello"
  }
  def foo = {
    object Hello extends Greeting
    Hello
  }
  import scala.language.existentials
  import scala.reflect.runtime.universe._
  def typeName[A: TypeTag](a: A) = {
    val t = typeOf[A]
    t.typeSymbol.toString
  }
  println("class is: " + foo.getClass)
  println("type is: " + typeName(foo))
}
```

This outputs:

```
class is: class example.Foo$Hello$1$
type is: <refinement of example.Foo.Greeting>
```

It's not possible for `Locate` here to deduce `example.Foo$Hello$1$.class` given `<refinement of example.Foo.Greeting>` or `<none>` or whatever.  Previously using `java.io.File` this type of query probably just returned `false`. I'm now doing it intentionally. Could we try to do the mapping between types and classes? Maybe but I am not sure if the added cost adds to the accuracy of https://github.com/sbt/zinc/issues/925 `isLibraryModified` in IncrementalCommon. Feel free to open a new issue if you think there something you want to look into.

Reply:
> It's not possible for `Locate` here to deduce `example.Foo$Hello$1$.class` given `<refinement of example.Foo.Greeting>` or `<none>` or whatever.

Yeah, and it shouldn't be `Locate`'s job to.  `<refinement of example.Foo.Greeting>` isn't a classname, so the root cause here isn't that Locate isn't checking for "<".

Do you know what introduced this regression?  What's the connection with `java.io.File`?

Reply:
I switched the internals to use more NIO Path, and that surfaced this behavior/bug. `java.io.File` would allow the creation of `<none>.class` whereas NIO Path on Windows would blow up saying invalid path on Windows.